File: 336-PalindromePairs/336PalindromePairsGuang.py
class Solution(object):
    def palindromePairs(self, words):
        def isPal(s):
            for i in xrange(len(s)/2):
                if s[i] != s[len(s)-1-i]:
                    return False
            return True
        wmap = {word : index for index, word in enumerate(words)}
        res = [];
        for idx, word in enumerate(words):
            if word != "" and isPal(word) and "" in wmap:
                res.append((wmap[""], idx))
            for i in xrange(len(word)):
                left, right = word[:i], word[i:]
                rl, rr = left[::-1], right[::-1]
                if isPal(right) and rl in wmap and wmap[rl] != idx:
                    res.append((idx, wmap[rl]))
                if isPal(left) and rr in wmap and wmap[rr] != idx:
                    res.append((wmap[rr], idx))
        return res            

File: 336-PalindromePairs/336PalindromePairsGuang.cpp
#define LL unsigned long long
class Solution {
public:
    LL prime = 17;
 
    LL hash(string s){
        LL p = 1;
        LL res = 0;
        for(auto c : s){
            res += (c-'a')*p;
            p *= prime;
        }
        return res;
    }
    LL rhash(string s){
        LL p = 1;
        LL res = 0;
        for(auto it = s.rbegin(); it != s.rend(); ++it){
            res += ((*it)-'a') * p;
            p *= prime;
        }
        return res;
    }
    unordered_map<string, pair<LL, LL>> hmap;
    vector<vector<int>> palindromePairs(vector<string>& words) {
        int n = words.size();        
        vector<vector<int>> res;
        // vector<LL> h, rh;
        // for(auto word : words){
        //     h.push_back(hash(word));
        //     rh.push_back(rhash(word));
        // }
        
        LL *h = new LL[n+4];
        LL *rh = new LL[n+4];
        int i = 0;
        for(auto word : words){
            h[i] = hash(word);
            rh[i] = rhash(word);
            ++i;
        }
        
        //vector<LL> p;
        LL p[100];
        LL t = 1;
        for(int i = 0; i < 100; i++){
            //p.push_back(t);
            p[i] = t;
            t *= prime;
        }
        for(int i = 0; i < n; i++){
            LL hi, hri, hj, hrj;
            hi = h[i], hri = rh[i];
            for(int j = i+1; j < n; j++){
                hj = h[j], hrj = rh[j];
                LL ij = hj*(LL)p[words[i].size()] + hi;
                LL ji = hi*(LL)p[words[j].size()] + hj;
                LL rjri = hri*(LL)p[words[j].size()] + hrj;
                LL rirj = hrj*(LL)p[words[i].size()] + hri;
                if(ij == rjri)// && isPal(words[i]+words[j]))
                    res.push_back(vector<int>({i, j}));
                if(ji == rirj)// && isPal(words[i]+words[j]))
                    res.push_back(vector<int>({j, i}));
            }
        }    
        return res;    
    }
};
